File: 2689.java
 import java.util.Scanner;

 public class Main {

     final static int MAXVALUE = (int) Math.sqrt(Integer.MAX_VALUE) + 1;

     public static void main(String[] args) {
         int k, l, u, tt, index;
         boolean hasFlg;
         int a, b, aa, bb;
         int[] num = new int[MAXVALUE];
         boolean[] frimeflg = new boolean[MAXVALUE];
         boolean[] flg;
         index = 1;
         // 鏋勯€犱竴涓皬绱犳暟琛?
         for (int i = 2; i < MAXVALUE; i++) {
             if (!frimeflg[i]) {
                 k = i * 2;
                 num[index++] = i;
                 while (k < MAXVALUE) {
                     frimeflg[k] = true;
                     k += i;
                 }
             }
         }
         Scanner scan = new Scanner(System.in);
         while (scan.hasNextInt()) {
             l = scan.nextInt();
             u = scan.nextInt();
             tt = (int) Math.sqrt(u);
             // 鏋勯€犱竴涓猽-l+1鐨勫尯闂?
             flg = new boolean[u - l + 1];
             // 绛涢€夊尯闂达紝涓嶆槸绱犳暟鐨勬爣璁颁负true
             for (int i = 1; i < index && i <= tt; i++) {
                 k = Math.max(num[i], l / num[i]);
                 if (k * num[i] < l || k <= 1) {
                     k++;
                 }
                 if (k <= 1) {
                     k++;
                 }
                 while ((long) k * num[i] <= u) {
                     flg[k * num[i] - l] = true;
                     k++;
                 }
             }
             hasFlg = false;
             a = 0;
             b = Integer.MAX_VALUE;
             aa = 0;
             bb = 0;
             if (l == 1) {
                 flg[0] = true;
             }
             for (int i = 0; i < u - l + 1; i++) {
                 if (!flg[i]) {
                     for (int j = i + 1; j <= u - l; j++) {
                         if (!flg[j]) {
                             if (b - a > j - i) {
                                 a = i + l;
                                 b = j + l;
                             }
                             if (bb - aa < j - i) {
                                 bb = j + l;
                                 aa = i + l;
                             }
                             hasFlg = true;
                             break;
                         }
                     }
                 }
             }
             if (hasFlg) {
                 System.out.printf(
                         "%d,%d are closest, %d,%d are most distant.\n", a, b,
                         aa, bb);
             } else {
                 System.out.println("There are no adjacent primes.");
             }
         }
     }
}

File: 2689.cc
#include <cstdio>
#include <cstring>
using namespace std;
int prime[4948],top=0;
bool isp[48005];
inline void init()
{
    int i,j,k;
    memset(isp,1,sizeof(isp));
    for(i=2;i<=48000;++i)
    {
        if(isp[i])prime[top++]=i;
        for(j=0;j<top;++j)
        {
            if(i*prime[j]>48000)break;
            isp[i*prime[j]]=0;
            if(!(i%prime[j]))break;
        }
    }
}
bool tmp[1000005];
inline void solve(int l,int r)
{
    int i,j,k,a,b,aa,bb;bool*p=tmp-l;
    int m=0x7fffffff,M=0x80000000;
    memset(tmp,1,sizeof(tmp));
    if(l==1)p[1]=0;
    for(i=0;i<top;++i)
    {
        if(prime[i]>r)break;
        j=(l-1)/prime[i]+1;if(j==1)++j;
        for(j*=prime[i];j>0&&j<=r;j+=prime[i])
            p[j]=0;
    }
    for(i=l;i<r;++i)
    {
        if(!p[i])continue;
        for(j=i+1;j<=r;++j)
            if(p[j])break;
        if(j>r)break;

        if(m>j-i){m=j-i;a=i;b=j;}
        if(M<j-i){M=j-i;aa=i;bb=j;}
    }
    if(m==0x7fffffff)
    {
        puts("There are no adjacent primes.");
        return;
    }
    printf("%d,%d are closest, %d,%d are most distant.\n",a,b,aa,bb);
}
int main()
{
    int l,r;
    init();
    while(~scanf("%d%d",&l,&r))
        solve(l,r);
    return 0;
}


